### ISSUE TITLE
Noticed list index out of range on ovs.vdisk.dtl_checkup in ovs_workers.log

### ISSUE THREAD

**Issue by pploegaert** [openvstorage/framework#777]
2016-08-05 11:36:56 35800 +0200 - ovs52n252 - 125179/140635645032256 - celery/celery.worker.job - 977 - ERROR - Task ovs.vdisk.dtl_checkup[61e2e1d9-cc37-40a7-b68d-0361671f41d7] raised unexpected: IndexError('list index out of range',)
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/celery/app/trace.py", line 240, in trace_task
    R = retval = fun(_args, *_kwargs)
  File "/usr/lib/python2.7/dist-packages/celery/app/trace.py", line 438, in **protected_call**
    return self.run(_args, *_kwargs)
  File "/opt/OpenvStorage/ovs/lib/helpers/decorators.py", line 236, in new_function
    output = function(_args, *_kwargs)
  File "/opt/OpenvStorage/ovs/lib/vdisk.py", line 956, in dtl_checkup
    current_sr = [sd for sd in vpool.storagedrivers if sd.storage_ip == current_dtl_config.host][0].storagerouter
IndexError: list index out of range

**Comment by matthiasdeblock** [comment]
Also, lot of these messages on GIG's env. (unstable)

```
2016-09-07 14:59:58 95300 +0200 - ovs-04.be-gen8-2 - 37168/140206635284288 - celery/celery.worker.job - 429 - ERROR - Task ovs.vdisk.dtl_checkup[faff60a2-27bb-438f-abc8-a0a501eac429] raised unexpected: IndexError('list index out of range',)
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/celery/app/trace.py", line 240, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/usr/lib/python2.7/dist-packages/celery/app/trace.py", line 438, in __protected_call__
    return self.run(*args, **kwargs)
  File "/opt/OpenvStorage/ovs/lib/helpers/decorators.py", line 236, in new_function
    output = function(*args, **kwargs)
  File "/opt/OpenvStorage/ovs/lib/vdisk.py", line 911, in dtl_checkup
    current_sr = [sd for sd in vpool.storagedrivers if sd.storage_ip == current_dtl_config.host][0].storagerouter
IndexError: list index out of range
```

**Comment by matthiasdeblock** [comment]
After creating a vpool on GIG's env. (same as mentioned earlier - unstable) we received the following as well.

```
2016-09-07 14:39:35 63900 +0200 - ovs-02.be-gen8-2 - 5025/139656891938624 - celery/celery.worker.job - 393 - ERROR - Task ovs.storagerouter.add_vpool[01ddc1d4-a86a-42d2-ac07-f1246e834f79] raised unexpected: IndexError('list index out of range',)
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/celery/app/trace.py", line 240, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/usr/lib/python2.7/dist-packages/celery/app/trace.py", line 438, in __protected_call__
    return self.run(*args, **kwargs)
  File "/opt/OpenvStorage/ovs/lib/storagerouter.py", line 1045, in add_vpool
    VDiskController.dtl_checkup(vpool_guid=vpool.guid, ensure_single_timeout=600)
  File "/usr/lib/python2.7/dist-packages/celery/local.py", line 167, in <lambda>
    __call__ = lambda x, *a, **kw: x._get_current_object()(*a, **kw)
  File "/usr/lib/python2.7/dist-packages/celery/app/trace.py", line 439, in __protected_call__
    return orig(self, *args, **kwargs)
  File "/usr/lib/python2.7/dist-packages/celery/app/task.py", line 420, in __call__
    return self.run(*args, **kwargs)
  File "/opt/OpenvStorage/ovs/lib/helpers/decorators.py", line 236, in new_function
    output = function(*args, **kwargs)
  File "/opt/OpenvStorage/ovs/lib/vdisk.py", line 911, in dtl_checkup
    current_sr = [sd for sd in vpool.storagedrivers if sd.storage_ip == current_dtl_config.host][0].storagerouter
IndexError: list index out of range
```

**Comment by wimpers** [comment]
There is a PR still open for this one https://github.com/openvstorage/framework/pull/863

**Comment by khenderick** [comment]
On @matthiasdeblock's environment, I noticed that the `dtl_config.host` for all disks was returning management ips (172.19.21.x) instead of storage ips (172.19.22.x). However, in the code configuring the dtl we only use the storage_ips.

**Comment by khenderick** [comment]
After some investigation, it seems that we ask the volumedriver to do automatic DTL configuration when a volume is created, and afterwards, we change it to a more appropriate location (taking things like domains into account). The volumedriver gets the ip address from its ClusterRegistry which the framework initially configures.

However, the framework can specify an identifier, an ip address and the management-, xmlrpc- and DTL port. It configures the management ip address since that's the one that should be used for management- and xmlrpc calls. Unfortunately, it can't specify a separate ip address for the DTL.

Possible solutions:
- The volumedriver allows us to configure a separate ip address for DTL that it can use for automatic configuration so the DTL uses the storage network as expected;
- We disable automatic configuration, but that can cause a (short) time in which no DTL will be configured;
- We ignore the fact that for this (short) time the DTL will run over a management network, and let the framework handle this ip mismatch and reconfigure the DTL correct after a (short) time.

**Comment by khenderick** [comment]
@wimpers, what solution from the above list is preferred?

**Comment by saelbrec** [comment]
@khenderick ,
First option sounds as the correct one, but option 3 could also be a possibility, but can you be more precise on "a (short) time" ?

@redlicha ,
For the "short time" that the DTL is configured over the management network it will impact performance - as that network might be bandwidth limited - or will it give up if it can't follow and go into a degraded state(which on its turn will trigger a dtl reconfiguration) ?

**Comment by khenderick** [comment]
@saelbrec, basically every time there is the need for reconfiguration (after a disk create event, after a disk migrate event, ...), we launch the dtl reconfiguration. However, due to celery's nature, there might be an undefined amount of time between the moment that the event is put on the queue and the moment it gets processed. Usually it will only take a matter of minutes, but obviously, it can't be exactly predicted.

**Comment by redlicha** [comment]
It will throttle, either immediately (synchronous mode) or once the client's buffer capacity is reached / a sync is requested (asynchronous mode).

FWIW, voldrv does not consider this a mgmt network - the traffic between voldrvs (redirected I/O) also uses that interface at the moment, so technically the XMLRPC one should be moved away.
While at it it makes however most sense to make it completely flexible.

WRT the alternatives listed by @khenderick:
- the circular setup (`Automatic` mode) maintained by voldrv has the downside that knowledge about the cluster topology (distant members) cannot be modelled and has the issues mentioned by @khenderick 
- @khenderick: option 2, where no DTL is configured for a short time: I suppose that's `Manual` mode with no DTL configured by default and then configuring a DTL per volume?
- there's a fourth option: use `Manual` mode and configure a default DTL per voldrv instance, override per volume as necessary (if at all). That way the DTL is set by voldrv immediately after creation / migration to the globally configured location. The catch with that one is that at the moment this default DTL cannot be reconfigured dynamically at the moment.

**Comment by khenderick** [comment]
Ticket openvstorage/volumedriver#113 has been resolved for further integration.

Implementation:
- (re)configure volumedrivers to have a circular manual DTL setup
- Adapt the code so it configures the correct information in the ClusterRegistry
- Fix the initial bug so the framework can handle unexpected ip addresses.

**Comment by khenderick** [comment]
After discussion with @saelbrec, the circular manual DTL will be postponed. Issue #993 will take this up.

**Comment by khenderick** [comment]
Fixed by #995, packaged in openvstorage-2.7.4-rev.4082.7183343

**Comment by JeffreyDevloo** [comment]
### Information
This issue has been covered by both a unittest ( ovs/lib/tests/vdisk_tests/test_dtl_checkup.py ).
